design/cluster-api-provider-metal3/configurable_disk_cleaning.md

```diff
@@ -0,0 +1,234 @@
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3MachineTemplate
+metadata:
+spec:
+  template:
+  spec:
+   disableAutomatedClean: False   #default value. Perform automated cleaning.
+```
+
+A new Metal3MachineTemplate controller will be introduced to reconcile
+Metal3MachineTemplate objects. When a user modifies `disableAutomatedClean`
+field on a Metal3MachineTemplate, CAPM3 Metal3MachineTemplate controller
+will update `disableAutomatedClean` field value on all the respective
+Metal3Machines referenced by that Metal3MachineTemplate.
+
+##### Metal3Machine object
+
+We introduce a new field `disableAutomatedClean` in the spec of the
+Metal3Machine object. `disableAutomatedClean` can store a boolean type
+of a value. Setting it to True means do not perform automated
+cleaning, while False value enables automated cleaning. By default
+`disableAutomatedClean` will be set to False (i.e. do perform automated
+cleaning as it is now).
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3Machine
+metadata:
+spec:
+  template:
+  spec:
+   disableAutomatedClean: False   #default value. Perform automated cleaning.
+
+```
+
+#### Baremetal Operator
+
+We introduce a new field `disableAutomatedClean` in the spec of the
+Baremetalhost object. `disableAutomatedClean` can store a boolean type
+of a value. If set to True, Baremetal Operator instructs Ironic
+to disable automated cleaning for an Ironic node, while if
+set to False, Baremetal Operator instructs Ironic to enable automated
+cleaning. By default `disableAutomatedClean` will be set to False
+(i.e. do perform automated cleaning as it is now).
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalHost
+metadata:
+  name: worker-0
+spec:
+  disableAutomatedClean: False   #default value. Perform automated cleaning.
+  ...
+```
+
+## Design Details
+
+**Example flow:**
+
+1. User wants to disable disk cleaning before upgrading the nodes that are part
+    of the same KCP/MD.
+
+2. User sets True for the `disableAutomatedClean` field on a
+    Metal3MachineTemplate which is referenced by `infrastructureTemplate` field
+    in the KCP/MD.
+
+3. Metal3MachineTemplate controller keeps reconciling the Metal3MachineTemplate objects.
+    Once the update is seen on the Metal3MachineTemplate, Metal3MachineTemplate controller
+    starts mapping all the Meta3Machines referenced by that particular Metla3MachineTemplate,
+    and updates the `disableAutomatedClean` field to True on all the referenced Metal3Machines.
+
+4. Metal3Machine controller keeps reconciling the Metal3Machine objects.
+    Once the update is seen on the Metal3Machine, Metal3Machine controller starts
+    mapping the BareMetalHosts referenced by the Metal3Machines, and updates the
+    `disableAutomatedClean` field to True on all the referenced BaremetalHosts.
+
+5. Since the `disableAutomatedClean` field is set to True on the BaremetalHost,
```

Comment:
Does it matter what state the host is in when this happens, either in the operator or in Ironic?

If the flag is false, should the baremetal-operator ensure that the corresponding flag in Ironic is also false?

Reply:
> Does it matter what state the host is in when this happens, either in the operator or in Ironic?

That's a good question. So, I would assume bmh will need to be either in `provisioned` or `ready` state, while an Ironic node is in `active` or `manageable`. I think I should mention this in the proposal right?

> If the flag is false, should the baremetal-operator ensure that the corresponding flag in Ironic is also false?

By default, an Ironic node will have `automated_clean` set to True/None If I'm not mistaken. And that would be equal to the default value (False) of the `disableAutomatedClean` field in bmh. And after that, whatever value is set, BMO will make sure to set the same in the Ironic node via the Ironic.

Reply:
Right, that was my point: We need to tell Ironic what value to use every time it is important for it to have the right value. How often do we need to do that? Just before deprovisioning? Every time we reconcile the host? Somewhere in between those two?

Reply:
TBH I'm not quite sure which option is better. For example, it make sense to update Ironic node just before the deprovisioning because upgrade will happen for already provisioned hosts, in other words before they go to deprovisioning. But on the other hand, maybe it is also good to update every time BMO reconciles the bmh so that we update the Ironic node asap.

Reply:
Also, how often we do depends on what BMH states we allow/recommend to update `disableAutomatedClean`. 

If we allow the user to directly update `disableAutomatedClean` field on a bmh, then BMO needs to make sure to update the Ironic node asap, or whenever reconciliation occurs. Because, the user can update the field when the bmh is in `Ready` state. So, we can't wait until deprovisioning happens, because the intention of the user here is to disable cleaning even during provisioning.
If we allow/recommend the user to directly update `disableAutomatedClean` field on a bmh, then BMO can update the Ironic node before going to deprovisioning. 

But now I think the better option would be to talk to Ironic to update node.automated_clean whenever BMO reconciles. We can also do the check in bmh to see if the current value of `disableAutomatedClean`  is the same or not. If not then, then it communicates with Ironic.

Reply:
I'm probably making a bigger deal out of this than necessary, but it's important to think clearly about how the value propagates through the system to ensure it has the right value when needed. We really only need to make sure we update the value before asking Ironic to do something that would rely on the value, right? What operations or state transitions in Ironic will have different behavior based on the setting? Maybe setting the value in Ironic only needs to be a step in the Deprovision() method of the provisioner?

Reply:
I replied to your first comment before reading the second.

If we also want to use this flag to disable cleaning during provisioning, then we would need to set it then, too.

I guess the point is, we don't need a step in our state machine or a separate reconciliation phase to set the value. If we want it done all the time, then we could do it in ValidateManagementAccess(). It seems reasonable to just set it for operations that would use it, though, and that might make it more obvious to someone trying to understand the code later.

Reply:
You are right that we don't need an extra step. I'm not fully sure yet where to add the logic for setting the value in Ironic. But I think the latter option you proposed makes more sense. But of course, I will have a better idea during the implementation.